examples/mixture_of_experts/dist_moe.py

```diff
@@ -0,0 +1,96 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates
+# torchrun --nproc-per-node 5 dist_moe.py
+
+import torch
+import torch.distributed as dist
+
+from pippy import annotate_split_points, pipeline, PipelineStage, SplitPoint
+from pippy.PipelineSchedule import PipelineScheduleGPipe
+
+
+d_hid = 16
+n_experts = 3
+batch_size = 4
+
+torch.manual_seed(0)
+
+# Each expert is a MLP
+class ExpertLayer(torch.nn.Module):
+    def __init__(self, d_hid):
+        super(ExpertLayer, self).__init__()
+        self.net1 = torch.nn.Linear(d_hid, d_hid)
+        self.relu = torch.nn.ReLU()
+        self.net2 = torch.nn.Linear(d_hid, d_hid)
+
+    def forward(self, x):
+        x = self.net1(x)
+        x = self.relu(x)
+        x = self.net2(x)
+        return x
+
+# Full model comprising n experts
+class MoE(torch.nn.Module):
+    def __init__(self, n_experts: int) -> None:
+        super().__init__()
+        self.pre_proc = torch.nn.Linear(d_hid, d_hid)
+        self.experts = torch.nn.ModuleList(
+            [
+                ExpertLayer(d_hid)
+                for _ in range(n_experts)
+            ]
+        )
+
+    def forward(self, x: torch.Tensor) -> torch.Tensor:
+        x = self.pre_proc(x)
+        outputs = []
+        for expert in self.experts:
+            outputs.append(expert(x))
+        return torch.cat(outputs, dim=1)
+
+
+dist.init_process_group()
+rank = dist.get_rank()
+world_size = dist.get_world_size()
+device = torch.device(f"cuda:{rank}")
+
+model = MoE(n_experts)
+x = torch.randn(batch_size, d_hid)
+
+# Mark the split point for each expert
+annotate_split_points(model, {f"pre_proc": SplitPoint.END})
+for i in range(n_experts):
+    annotate_split_points(
+        model, {f"experts.{i}": SplitPoint.END}
+    )
+
+pippy_model = pipeline(model, 1, (x,))
+
+assert pippy_model.num_stages == world_size
+if rank == 0:
+    print("Original model:\n", model)
+    print("PiPPy model:")
+    pippy_model.print_readable()
```

Comment:
I got an error on this line.  Added a debug print for type:
`type <class 'pippy.IR.Pipe'>
PiPPy model equivalent: 2.4520821571350098 ref 2.4520821571350098
[rank0]: Traceback (most recent call last):
[rank0]:   File "/data/users/less/local/pippy/examples/moe/moe.py", line 74, in <module>
[rank0]:     pippy_model.print_readable()
[rank0]:   File "/home/less/local/miniconda3/envs/gocuda/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1709, in __getattr__
[rank0]:     raise AttributeError(f"'{type(self).__name__}' object has no attribute '{name}'")
[rank0]: AttributeError: 'Pipe' object has no attribute 'print_readable'`

Reply:
Sorry, the `print_readable()` method is in a pending PR: https://github.com/pytorch/PiPPy/pull/1037